Add `pushd`, `popd` and `dirs` commands for iocsh

JJL772 opened the pull request:
These are simplified versions of the bash equivalents detailed here: https://www.gnu.org/software/bash/manual/html_node/Directory-Stack-Builtins.html

`pushd` is used like `pushd path/to/my/directory`, and takes no other arguments or flags.

`popd` takes no arguments and removes the top item from the directory stack.

`dirs` shows the entire directory stack with the current working directory on top. Even if the stack is empty, `dirs` will still display the cwd.

I also used struct hack in the `dirStackEntry` structure. This is standard in C99, but was provided as a compiler extension prior to then. GCC as far back as 3.4.6 (at least) seems to [support this without issue](https://godbolt.org/z/94jEMTsd6). That being said, I don't know what support for these is like in Visual Studio. Everything that I can find seems to indicate that it's been supported since at least 2012. We'll see what Appveyor says :) 
According to the Microsoft docs, flexible arrays are available as long as Microsoft extensions are enabled (i.e. `/permissive-` and `/Za` are *not* set).

Closes #494

mdavidsaver commented on modules/libcom/src/iocsh/libComRegister.c:
```diff
@@ -28,6 +29,13 @@
+	char path[];
+};
+
+static ELLLIST dirStack = ELLLIST_INIT;
```
IOC shell functions may run concurrently.  For basic user sanity, I don't think it is reasonably to have all threads share a common stack.

I think it would be better to add this stack to the per-thread `struct iocshContext`.

mdavidsaver replied:
Also, since `iocsh.cpp` is C++, maybe `std::list<std::string>` would avoid some effort, and avoid memory leaks?

anjohnson replied:
Do individual Posix, Windows and RTEMS-score threads each have their own record of the CWD, or do they share one CWD for the whole process? I'm pretty sure VxWorks only has one CWD, although I'm be okay ignoring that target if it's unique. If it isn't and concurrent iocsh instances are calling `cd`/`pushd`/`popd` there are going to be some major problems with relative paths. We tell people never to use relative paths for the access security configuration file because someone running a `cd` command would cause a DoS if/when that file gets reloaded (iocStats loads a PV for doing that).

mdavidsaver replied:
@anjohnson I am not referring to `getcwd()`, but only to this iocsh directory stack.  I acknowledge that it is not possible to avoid races with concurrent `chdir()`.  I understand though that one of the goals of this change is to bring a measure of sanity.  I think that a per-thread stack will help with this.

JJL772 replied:
I've added this to `iocshContext` and switched to a `std::list` of `std::string`s

mdavidsaver commented on modules/libcom/src/iocsh/libComRegister.c:
```diff
@@ -470,6 +478,65 @@ static void installLastResortEventProviderCallFunc(const iocshArgBuf *args)
+	struct dirStackEntry* e = NULL;
+
+	char cwd[PATH_MAX];
+	getcwd(cwd, sizeof(cwd));
```
You need to handle the possibility of failure here.  eg. if the current directory name is very long.  Also, `PATH_MAX` may not be available on all targets.

Actually, it looks like windows supports the posix-2001 behavior for `getcwd(NULL,0)` to return a `malloc()` buffer.  So maybe something like this untested code:

```c++
std::string cwd;
#ifdef PATH_MAX
char scratch[PATH_MAX];
if(getcwd(scratch, sizeof(scratch))) {
    cwd = scratch;
} else
#endif
if(char *alloc = getcwd(NULL,0)) {
    try {
        cwd = alloc;
        free(alloc);
    } catch(std::exception&) { // only bad_alloc possible
        free(alloc);
    }
}
if(!cwd.empty())
    context->dirstack.push(cwd);
else
    iocshSetErr(1);
```

JJL772 replied:
It might be worth investigating other uses of `getcwd` as well. It seems like `PATH_MAX` is defaulted to 100 in a couple of source files, even though Windows provides `MAX_PATH` instead of `PATH_MAX`. The `cd` command uses a seemingly arbitrary buffer size of 1024 with `getcwd`.

`PATH_MAX` is defined in `limits.h` (part of the C library), and it exists in both newlib and glibc. Windows has its own `MAX_PATH` and doesn't define `PATH_MAX`. Perhaps EPICS should provide `EPICS_PATH_MAX` to consolidate these definitions and to provide a default if it's missing?

JJL772 replied:
I've added error handling to this particular call of `getcwd`, but haven't used the POSIX 2001 behavior for it. I'd like to get some feedback on the above comment before I go one way or the other.

mdavidsaver replied:
> It might be worth investigating other uses of `getcwd` ...

I agree, though imo. this should be a separate project.

Since these commands are a new feature, I think I am ok with not handling the long CWD case right away.

mdavidsaver commented on modules/libcom/src/iocsh/iocsh.cpp:
```diff
@@ -1542,6 +1560,85 @@ static void onCallFunc(const iocshArgBuf *args)
+        return;
+    }
+
+    if (iocshSetError(chdir(args[0].sval)) != 0) {
```
fyi. all iocsh arguments are optional.  So `args[0].sval` can be `NULL`.

mdavidsaver commented on modules/libcom/src/iocsh/iocsh.cpp:
```diff
@@ -1542,6 +1560,85 @@ static void onCallFunc(const iocshArgBuf *args)
+    }
+
+    if (iocshSetError(chdir(ctxt->dirStack.front().c_str())) != 0) {
+        fprintf(stderr, "Invalid directory path, ignored\n");
```
This error isn't be ignored.  (nor should it be)  Please correct the message.

mdavidsaver commented on modules/libcom/src/iocsh/iocsh.cpp:
```diff
@@ -1542,6 +1560,85 @@ static void onCallFunc(const iocshArgBuf *args)
+    if (!getcwd(cwd, sizeof(cwd))) {
+        iocshSetError(errno);
+        perror("getcwd failed");
+        return;
```
Minor, but the dirStack could still be printed in this situation.

mdavidsaver commented on modules/libcom/src/iocsh/iocsh.cpp:
```diff
@@ -1084,10 +1102,10 @@ iocshBody (const char *pathname, const char *commandLine, const char *macros)
 
     if (!context) {
-        context = (iocshContext*)calloc(1, sizeof(*context));
+        context = new iocshContext;
```
Use [`std::nothrow`](https://en.cppreference.com/w/cpp/memory/new/nothrow) here.

```c++
context = new(std::nothrow) iocshContext;
```

mdavidsaver reviewed (changes requested):
Looking better.

JJL772 commented:
I'd like #510 to be merged first so I can use `PATH_MAX`

mdavidsaver commented on modules/libcom/src/iocsh/iocsh.cpp:
```diff
@@ -1545,6 +1563,94 @@ static void onCallFunc(const iocshArgBuf *args)
+{
+    if (!args[0].sval) {
+        iocshSetError(-1);
+        fprintf(stderr, "Invalid directory path\n");
```
Maybe a message would be: `"Missing required argument: dir"`?

mdavidsaver commented:
> I'd like #510 to be merged first so I can use `PATH_MAX`

@JJL772 I'm switching this PR to Draft pending development of an alternative to #510.

JJL772 committed:
Add epicsGetCwd() function

This returns a heap allocated buffer containing the current working
directory. Slightly more portable than using getcwd() directly.

JJL772 committed:
Switch getcwd calls to epicsGetCwd

JJL772 committed:
Add pushd, popd and dirs commands to iocsh

These roughly mirror their bash built-in equivalents. Output
formatting is slightly different.
